birdhouse/optional-components/prometheus-longterm-metrics/default.env

```diff
@@ -0,0 +1,30 @@
+export PROMETHEUS_LONGTERM_VERSION='${PROMETHEUS_VERSION:-"v2.52.0"}'
+export PROMETHEUS_LONGTERM_DOCKER='${PROMETHEUS_DOCKER:-prom/prometheus}'
+export PROMETHEUS_LONGTERM_IMAGE='${PROMETHEUS_LONGTERM_DOCKER}:${PROMETHEUS_LONGTERM_VERSION}'
+
+export PROMETHEUS_LONGTERM_RETENTION_TIME=1y
+export PROMETHEUS_LONGTERM_SCRAPE_INTERVAL=1h
+
+# These are the prometheus defaults
+export PROMETHEUS_LONGTERM_TSDB_MIN_BLOCK_DURATION=2h
+export PROMETHEUS_LONGTERM_TSDB_MAX_BLOCK_DURATION=1d12h
+
+# These are the targets that
+export PROMETHEUS_LONGTERM_TARGETS='["prometheus:9090"]' # yaml list syntax
+
+OPTIONAL_VARS="
+  $OPTIONAL_VARS
+  \$PROMETHEUS_LONGTERM_SCRAPE_INTERVAL
+  \$PROMETHEUS_LONGTERM_TARGETS
+"
+
+export DELAYED_EVAL="
+  $DELAYED_EVAL
+  PROMETHEUS_LONGTERM_VERSION
+  PROMETHEUS_LONGTERM_DOCKER
+  PROMETHEUS_LONGTERM_IMAGE
+  PROMETHEUS_LONGTERM_RULES_FILE
+"
+
+# Note that this component does not depend explicitly on the `components/monitoring` component so that this can
+# theoretically be deployed on a different machine than the `prometheus` service. This is currently untested.
```

Comment:
Curious how that would work out?
It seems to depend on `prometheus:9090` and extends the volumes of `prometheus` service. 
I think it is more confusing to omit the dependency.

Reply:
It's actually a completely separate service with no shared volumes. It does depend on `prometheus:9090` by default but that default can be updated by setting the `PROMETHEUS_LONGTERM_TARGETS` variable.

`PROMETHEUS_LONGTERM_TARGETS` can point to any prometheus endpoints at all. 

See the discussion about this here: https://github.com/bird-house/birdhouse-deploy/pull/461#discussion_r1644964487

Reply:
What I meant is that the rules are mounted under `prometheus` service here: 
https://github.com/bird-house/birdhouse-deploy/blob/longterm-monitoring/birdhouse/optional-components/prometheus-longterm-rules/config/monitoring/docker-compose-extra.yml

So, whether `PROMETHEUS_LONGTERM_TARGETS` refers to the same instance or a custom remote one, the local instance needs to have the long-term rules defined. Therefore, doesn't that mean that `prometheus` service (i.e.: using `components/monitoring`) becomes mandatory regardless, since it is needed to mount the rules?

Reply:
Yes, prometheus must be running on the same host as the rest of the birdhouse stack. 

That first prometheus instance must have at least one rule that has the `longterm-metrics` group label in order for a second prometheus instance (either running on the same machine or elsewhere) to know which metrics it should monitor and store from the first prometheus instance.

Reply:
I've taken a second look at the directory structure, and starting to see why `components/monitoring` is not a dependency. Is it only because `components/monitoring` combines cardvisor, grafana, alertmanager, prometheus, etc. all together?

Therefore, if someone wants "only" longterm-metrics, they need to define their own `prometheus` service with a definition similar to this?
https://github.com/bird-house/birdhouse-deploy/blob/3f75d496b932ab4d90857206003d0225cd20c435/birdhouse/components/monitoring/docker-compose-extra.yml#L32-L49

I think part of what makes all of this confusing is that there are 2 different, though related, component definitions (and I still do not understand why): 
- [optional-components/prometheus-longterm-metrics](https://github.com/bird-house/birdhouse-deploy/tree/longterm-monitoring/birdhouse/optional-components/prometheus-longterm-metrics)
- [optional-components/prometheus-longterm-rules](https://github.com/bird-house/birdhouse-deploy/tree/longterm-monitoring/birdhouse/optional-components/prometheus-longterm-rules)

From a user's perspective, this all seems really convoluted.
Isn't there are way to simplify this hierarchy? For example, what if there was a distinct `optional-components/prometheus` definition, and `components/monitoring` depends on it. Then, [optional-components/prometheus-longterm-metrics](https://github.com/bird-house/birdhouse-deploy/tree/longterm-monitoring/birdhouse/optional-components/prometheus-longterm-metrics) could also depend only on `optional-components/prometheus`?

Reply:
@fmigneault 

> Therefore, if someone wants "only" longterm-metrics, they need to define their own prometheus service with a definition similar to this?

If someone **only** wants longterm metrics then they can have a single prometheus service that sets `--storage.tsdb.retention.time` to something longer than 90 days.

However, if you want some metrics to be stored for longer than others, you need a second prometheus instance. This is because the `--storage.tsdb.retention.time` value is set for an entire prometheus instance (see discussion here: https://github.com/bird-house/birdhouse-deploy/issues/277#issuecomment-2043273300 and here: https://github.com/bird-house/birdhouse-deploy/pull/461#discussion_r1644986997).

> I think part of what makes all of this confusing is that there are 2 different, though related, component definitions (and I still do not understand why)

These are separate in order to accommodate @tlvu's request to be able to deploy the `prometheus-longterm-metrics` component on a different server. `[optional-components/prometheus-longterm-rules](https://github.com/bird-house/birdhouse-deploy/tree/longterm-monitoring/birdhouse/optional-components/prometheus-longterm-rules) are just suggested default rules to be added to the `prometheus` server. As described in the README file:

```
This adds some default longterm metrics rules to the `prometheus` component 
```

These rules are added to the `prometheus` component on the same machine that is running the rest of the stack because that is where all the relevant metrics are collected by the rest of the monitoring components (cadvisor, nodeexporter). These rules are not necessary, they are recommended defaults but you can customize them however you like. It is easier to customize and add your own if the rules are separate. Users can choose to include the defaults, or not.

Reply:
Agree with Misha's explanations.  I just want to add that each organization needs are different so we should not force a certain "configuration".

Specifically to this PR, we should not force the 2 Prometheus to be on the same machine.  So `prometheus-longterm-metrics` should not depend on `components/monitoring`.  We should also not be forced to use `optional-components/prometheus-longterm-rules` because each org will probably have different metrics they deemed useful to keep for longterm.

During Misha's leave, a sysadmin from PCIC actually had a question about how to pull existing metrics from our PAVICS Prometheus into his own centralized Prometheus for a centralized view of all his servers in one place.  I point him to this PR for inspiration.

Reply:
My bad. I was not clear in my explanation. What I intended to say was:
_"if someone wants only `prometheus` monitoring **with both long-term and short term metrics**, they need to define their own `prometheus` service_" [...], because they cannot activate `prometheus` on its own (i.e.: they must use the full package `components/monitoring` that includes all other services).

Having short/long-term separate, and having them configurable on different servers is perfectly fine by me. I'm all for that flexibility, and never mentioned otherwise.

It just seems from the configuration files that, in order for long-term metrics to be sent to the remote server, the *rules* must be mounted into the short-term local `prometheus` (as shown below), and since that `prometheus` must exist, then it should be a dependency. I understand `long-term-prometheus` NOT being a dependency to toggle it independently of the short-term one. This is good. Here I am referring to https://github.com/bird-house/birdhouse-deploy/tree/3f75d496b932ab4d90857206003d0225cd20c435/birdhouse/optional-components/prometheus-longterm-rules having `component/monitoring` dependency, NOT https://github.com/bird-house/birdhouse-deploy/tree/3f75d496b932ab4d90857206003d0225cd20c435/birdhouse/optional-components/prometheus-longterm-metrics. In other words, if one adds `optional-components/prometheus-longterm-rules` **ONLY** without also thinking to add `components/monitoring`, docker-compose will complain that `service: prometheus` is not defined.

https://github.com/bird-house/birdhouse-deploy/blob/3f75d496b932ab4d90857206003d0225cd20c435/birdhouse/optional-components/prometheus-longterm-rules/config/monitoring/docker-compose-extra.yml#L3-L6

Reply:
Ok I understand your concern now.

> In other words, if one adds optional-components/prometheus-longterm-rules ONLY without also thinking to add components/monitoring, docker-compose will complain that service: prometheus is not defined.

Docker compose won't actually complain in this case. Additional settings defined under the `config/` folder in components are only applied if the component that it references is also enabled. So in this case, if you add `prometheus-longterm-rules` as a component but not `monitoring`, then nothing will happen (since the additional docker-compose settings are in `prometheus-longterm-rules/config/monitoring`.

If we do make the `prometheus-longterm-rules` component dependent on the `monitoring` components then if someone adds the `prometheus-longterm-rules` but **not** the `monitoring` rule to the `BIRDHOUSE_EXTRA_CONF_DIRS` configuration variable, then `monitoring` **will** be added in automatically. This is also a surprising behaviour that users may not expect.

My opinion is that we should leave it as is, so that users must specify both components if they want both components. I'm open to other opinions on this subject though.

Reply:
Thanks for validating how this resolves.
I'm fine with either approach if the README indicates clearly what is supposed to happen and what to do if the one or the other situation is desired.